packages/nextjs/src/server/clerkMiddleware.ts

```diff
@@ -178,11 +201,33 @@ export const clerkMiddleware = ((...args: unknown[]): NextMiddleware | NextMiddl
       } catch (e: any) {
         handlerResult = handleControlFlowErrors(e, clerkRequest, request, requestState);
       }
+      if (options.contentSecurityPolicy) {
+        const { header, nonce } = createCSPHeader(
+          options.contentSecurityPolicy.mode,
+          (parsePublishableKey(publishableKey)?.frontendApi ?? '').replace('$', ''),
+          options.contentSecurityPolicy.directives,
+        );
+
+        setHeader(handlerResult, 'Content-Security-Policy', header);
+        if (nonce) {
+          setHeader(handlerResult, 'X-Nonce', nonce);
+        }
+
+        logger.debug('Clerk generated CSP', () => ({
+          header,
+          nonce,
+        }));
+      }
 
       // TODO @nikos: we need to make this more generic
       // and move the logic in clerk/backend
       if (requestState.headers) {
         requestState.headers.forEach((value, key) => {
+          if (key === 'Content-Security-Policy') {
```

Comment:
Minor: we can replace this string this with the `constants` import from `@clerk/backend`